packages/blocks/src/_common/components/block-component.ts

```diff
@@ -0,0 +1,43 @@
+import { BlockElement, type BlockService } from '@blocksuite/block-std';
+import type { BlockModel } from '@blocksuite/store';
+import { html, nothing } from 'lit';
+import { query } from 'lit/decorators.js';
+import { type StyleInfo, styleMap } from 'lit/directives/style-map.js';
+
+import type { BlockCaptionEditor } from './block-caption.js';
+
+export class BlockComponent<
+  Model extends BlockModel = BlockModel,
+  Service extends BlockService = BlockService,
+  WidgetName extends string = string,
+> extends BlockElement<Model, Service, WidgetName> {
+  protected accessor useCaptionEditor = false;
+
+  protected accessor blockContainerStyles: StyleInfo | undefined = undefined;
+
+  @query('.affine-block-component > block-caption-editor')
+  private accessor _captionEditor!: BlockCaptionEditor | null;
+
+  get captionEditor() {
+    if (!this.useCaptionEditor || !this._captionEditor)
+      throw new Error(
+        'Oops! Please enable useCaptionEditor before accessing captionEditor'
+      );
+    return this._captionEditor;
+  }
+
+  override render() {
```

Comment:
Instead of override `render`, you should override `renderBlock`

Reply:
But the subclass will override the `renderBlock` method.

Reply:
The problem of this code is that, it will always try to render `affine-block-selection` even if the block is [in `Hide` or `Bypass` mode](https://github.com/toeverything/blocksuite/blob/master/packages/framework/block-std/src/view/element/block-element.ts#L301C5-L320C8). And even wroth, when the component is in error mode, thoses components will still be rendered.

Reply:
> The problem of this code is that, it will always try to render `affine-block-selection` even if the block is [in `Hide` or `Bypass` mode](https://github.com/toeverything/blocksuite/blob/master/packages/framework/block-std/src/view/element/block-element.ts#L301C5-L320C8). And even wroth, when the component is in error mode, thoses components will still be rendered.

Yeah, I noticed it. PTAL of the new version.

Reply:
This will be hard to maintain if we add new view types. I'll add a new solution here in your branch.

Reply:
Done